File: 2220-find-all-possible-recipes-from-given-supplies/find-all-possible-recipes-from-given-supplies.java
class Solution {
    public List<String> findAllRecipes(String[] recipes, List<List<String>> ingredients, String[] supplies) {
        Map<String, List<String>> graph = new HashMap<>();
        Map<String, Integer> inDegree = new HashMap<>();

        for (String recipe : recipes) {
            inDegree.put(recipe, 0);
        }
        
        for (int i = 0; i < recipes.length; i++) {
            String recipe = recipes[i];
            List<String> ingredientList = ingredients.get(i);

            for (String ingredient : ingredientList) {
                graph.computeIfAbsent(ingredient, x -> new ArrayList<>()).add(recipe);
                inDegree.put(recipe, inDegree.get(recipe) + 1);
            } 
        }

        Queue<String> queue = new LinkedList<>();
        Set<String> available = new HashSet<>(Arrays.asList(supplies));

        for (String supply : supplies) {
            queue.offer(supply);
        }

        List<String> result = new ArrayList<>();

        while (!queue.isEmpty()) {
            String current = queue.poll();

            if (inDegree.containsKey(current)) {
                result.add(current);
            }

            if (graph.containsKey(current)) {
                for (String dependentRecipe: graph.get(current)) {
                    inDegree.put(dependentRecipe, inDegree.get(dependentRecipe) - 1);

                    if (inDegree.get(dependentRecipe) == 0) {
                        queue.offer(dependentRecipe);
                    }
                }
            }
        }

        return result;
    }
}

File: 2220-find-all-possible-recipes-from-given-supplies/find-all-possible-recipes-from-given-supplies.go
func findAllRecipes(recipes []string, ingredients [][]string, supplies []string) []string {
    graph := make(map[string][]string);
    inDegree := make(map[string]int);

    for _, recipe := range recipes {
        inDegree[recipe] = 0;
    }

    for i, recipe := range recipes {
        for _, ing := range ingredients[i] {
            graph[ing] = append(graph[ing], recipe);
            inDegree[recipe]++;
        }
    }

    queue := make([]string, 0);
    available := make(map[string]bool);

    for _, supply := range supplies {
        queue = append(queue, supply);
        available[supply] = true;
    }

    result := make([]string, 0);

    for len(queue) > 0 {
        current := queue[0];
        queue = queue[1:];

        if _, exists := inDegree[current]; exists {
            result = append(result, current);
        }

        for _, dependentRecipe := range graph[current] {
            inDegree[dependentRecipe]--;

            if inDegree[dependentRecipe] == 0 {
                queue = append(queue, dependentRecipe);
            }
        }
    }

    return result;
}